backend/mora/auth/sso.py

```diff
@@ -5,38 +5,176 @@
 # License, v. 2.0. If a copy of the MPL was not distributed with this
 # file, You can obtain one at http://mozilla.org/MPL/2.0/.
 #
-
+import functools
 import os
+from urllib import parse
 
 import flask
-import flask_saml
+from onelogin.saml2.auth import OneLogin_Saml2_Auth
+from onelogin.saml2.idp_metadata_parser import OneLogin_Saml2_IdPMetadataParser
+from onelogin.saml2.utils import OneLogin_Saml2_Utils
 
 from . import base
-from . import tokens
-from .. import settings
 
 basedir = os.path.dirname(__file__)
 
+blueprint = flask.Blueprint('sso', __name__, static_url_path='',
+                            url_prefix='/saml')
+
+
+def get_saml_settings(app):
+    config = app.config
+
+    insecure = config['SAML_IDP_INSECURE']
+    cert_file = config['SAML_CERT_FILE']
+    key_file = config['SAML_KEY_FILE']
+    requests_signed = config['SAML_REQUESTS_SIGNED']
+    saml_idp_metadata_url = config['SAML_IDP_METADATA_URL']
+
+    remote = OneLogin_Saml2_IdPMetadataParser.parse_remote(
+        saml_idp_metadata_url,
+        validate_cert=not insecure
+    )
+
+    s = {
+        "strict": True,
+        "debug": True,
+        "sp": {
+            "entityId": flask.url_for('sso.metadata', _external=True),
+            "assertionConsumerService": {
+                "url": flask.url_for('sso.acs', _external=True),
+                "binding": "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST"
+            },
+            "singleLogoutService": {
+                "url": flask.url_for('sso.sls', _external=True),
+                "binding": "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect"
+            },
+        },
+        "security": {
+            "authnRequestsSigned": requests_signed,
+            "logoutRequestSigned": requests_signed
+        }
+    }
+
+    s.setdefault('sp', {}).update(remote.get('sp'))
+    s.setdefault('idp', {}).update(remote.get('idp'))
+
+    if requests_signed:
+        with open(cert_file, 'r') as cf:
+            cert = cf.read()
+        with open(key_file, 'r') as kf:
+            key = kf.read()
+
+        s['sp'].update({
+            "x509cert": cert,
+            "privateKey": key
+        })
+
+    return s
+
+
+def prepare_flask_request():
+    # If server is behind proxys or balancers use the HTTP_X_FORWARDED fields
+    url_data = parse.urlparse(flask.request.url)
+    return {
+        'https': 'on' if flask.request.scheme == 'https' else 'off',
+        'http_host': flask.request.host,
+        'server_port': url_data.port,
+        'script_name': flask.request.path,
+        'get_data': flask.request.args.copy(),
+        'post_data': flask.request.form.copy()
+    }
+
+
+def init_saml_auth(func):
```

Comment:
What does this do? Also, is `init` a good prefix?

Reply:
It creates and initializes a SAML client and passes it along to whatever function this is annotated by.
``init_saml_auth`` is probably not the best name. I'll think of something better.